File: backend/internal/validation/phone.go
// Package validation holds small, reusable field-format checks shared across
// services — not a general request-validation framework, just the handful of
// formats (starting with phone numbers) that need the same rule in more than
// one place.
package validation

import (
	"errors"
	"regexp"
)

// ErrInvalidPhone is returned by services when a phone number is present but
// doesn't match the Sri Lankan format.
var ErrInvalidPhone = errors.New("phone number must be a valid Sri Lankan number (e.g. 0771234567, 0112345678, or +94771234567)")

// sriLankanPhone matches a leading +94 or 0 followed by exactly 9 digits —
// the same shape covers both mobile numbers (07x...) and landlines with a
// 2-digit area code (011, 021, ... ), since Sri Lankan subscriber numbers are
// always a 2-digit prefix + 7 digits after the leading 0/+94.
var sriLankanPhone = regexp.MustCompile(`^(?:\+94|0)\d{9}$`)

// IsValidSriLankanPhone reports whether s is a validly formatted Sri Lankan
// phone number. An empty string is considered valid — most phone fields in
// this app are optional, so callers combine this with their own presence
// check wherever a phone number is actually required.
func IsValidSriLankanPhone(s string) bool {
	if s == "" {
		return true
	}
	return sriLankanPhone.MatchString(s)
}

File: frontend/src/lib/phone.ts
// Matches a leading +94 or 0 followed by exactly 9 digits — the same shape
// covers both mobile numbers (07x...) and landlines with a 2-digit area code
// (011, 021, ...), since Sri Lankan subscriber numbers are always a 2-digit
// prefix + 7 digits after the leading 0/+94. Mirrors the backend's
// internal/validation/phone.go — keep both in sync if this ever changes.
const SRI_LANKAN_PHONE = /^(?:\+94|0)\d{9}$/;

// An empty string is considered valid here — most phone fields are optional,
// so callers combine this with their own presence check where a phone number
// is actually required.
export function isValidSriLankanPhone(value: string): boolean {
  const trimmed = value.trim();
  if (trimmed === "") return true;
  return SRI_LANKAN_PHONE.test(trimmed);
}

export const PHONE_INVALID_TEXT =
  "Enter a valid Sri Lankan number, e.g. 0771234567, 0112345678, or +94771234567.";
